### DISCUSSION TITLE
Date Picker - pickadate

### DISCUSSION THREAD

**Discussion by brouwersron** [livewire/livewire#3902] [1. Help]
So I am trying to get this datepicker (https://amsul.ca/pickadate.js/date/) to work with Laravel Livewire. Basically what I have so far is the following component:

<div wire:ignore
    {{ $attributes }}
    x-data
    @onSet="$dispatch('input', $event.target.value)"
>

    <div class="form-group form-group-feedback form-group-feedback-left">

        <input type="text" class="form-control datepicker">

    </div>

</div>

And in my form I call it like this:

    <x.input.datepicker wire:model="date"></x.input.datepicker>

It works as in, when I select a date and then change another field, the date is still selected and is not being removed. When I submit the form however, the field is empty, so the value isn't actually set. My problem I think is here:

    @onSet="$dispatch('input', $event.target.value)"

I took the above from one of the screencasts, but my knowledge about alpine and javascript in general is not so great. The "onSet" is an event from the datepicker itself (https://amsul.ca/pickadate.js/date/#events).

So my guess is, this should be an easy javascript solution perhaps. Like, "onchange, set the date".

**Comment by bobwurtz** [comment]
You want to use `Livewire.emit` from the global JS. See the docs [here](https://laravel-livewire.com/docs/2.x/events#firing-events).

I think that the `$dispatch` you are mentioning is an incorrect [magic action](https://laravel-livewire.com/docs/2.x/actions#magic-actions) that does not actually exist, unless I'm missing something. I believe that magic actions only work on user actions (like a click) where you can directly set that attribute in the element. But since you are using a datepicker I don't think that will work here.

So, based on the sample Datepicker code you linked to, I think you want something like this:

```
$('.datepicker').pickadate({
  onStart: function() {
    console.log('Hello there :)')
  },
  onRender: function() {
    console.log('Whoa.. rendered anew')
  },
  onOpen: function() {
    console.log('Opened up')
  },
  onClose: function() {
    console.log('Closed now')
  },
  onStop: function() {
    console.log('See ya.')
  },
  onSet: function(context) {
      var theSelectedDate = $('#date).val();  // idk if context holds the date value or how that works but get the value of the date
      Livewire.emit('setDate', theSelectedDate);  // then pass it as the second value in this function call
  }
})
```

You will need to add a [listener](https://laravel-livewire.com/docs/2.x/events#event-listeners) to the backend Livewire component. But then you should be able to set the date value to whatever you collect with the JS.

**Reply by joshhanley** [reply]
@bobwurtz `$dispatch` is an Alpine helper method, so will work fine. https://alpinejs.dev/magics/dispatch

But the issue is `onSet` isn't a browser event, it's just an internal "event" name for the library.

I think using `emit` and a backend listener is probably a bit much, have a look at my answer below where you can either dispatch an input event or make use of `@this` to set properties directly, which is what `wire:model` does under the hood.

Hope this helps!

**Comment by joshhanley** [comment]
@brouwersron the issue is that `onSet` is actually just a configuration property on the pickadate object, not a dispatched browser event.

But you should be able to dispatch an input event from within the `onSet` function call back, which Livewire can listen for, if you want to make it a reusable component.

If you are happy instantiating each datepicker though, a simpler way is to actually just set your Livewire property from within your `onSet` function. So if you had a property called `$date` on your Livewire component, you can do something like

```js
onSet: function(context) {
    @this.set('date', context) // @this get's the current Livewire components javascript instance. This will only work if `context` is the selected date but you should be able to console log it to confirm.
}
```

See here for more details on `@this` https://laravel-livewire.com/docs/2.x/inline-scripts

Hope this helps!
